### PR TITLE
fix(HLS): Fix missing roles

### LINKED ISSUE #4759: HLS subtitles lose role information
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
yes

**What version of Shaka Player are you using?**
latest

**Can you reproduce the issue with our latest release version?**
yes

**Can you reproduce the issue with the latest code from `main`?**
yes

**Are you using the demo app or your own custom app?**
own custom app

**If custom app, can you reproduce the issue using our demo app?**
n/a

**What browser and OS are you using?**
Chrome, macOS

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**

**What are the manifest and license server URIs?**

n/a

**What configuration are you using?  What is the output of `player.getConfiguration()`?**

n/a

**What did you do?**

Play HLS content that has both captions and subtitles, the captions having the EXT-MEDIA-TAG as follows:
```
#EXT-X-MEDIA:TYPE=SUBTITLES,GROUP-ID="sub1",LANGUAGE="sv",NAME="Swedish [SDH]",AUTOSELECT=YES,CHARACTERISTICS="public.accessibility.describes-music-and-sound",URI="sc-gaFESg/s2_index.m3u8"
```

**What did you expect to happen?**
The subtitle track parsed as having `roles: ["public.accessibility.describes-music-and-sound"]`

**What actually happened?**
There is no information in the shaka track object regarding what `roles` (or `kind`) this track has, so there is no way of distinguishing it as captions.

I have a fix locally that solves the problem, and could make a pull request. It would be good to have someone with more knowledge of `hls_parser.js` to take a look at it.

### PR DESCRIPTION
**Pull request by fredrik-telia**
Issue wasn't caught by the existing tests.

Closes #4759

### CODE REVIEW

**Comment by avelad** [comment]
@fredrik-telia can you add a regression test? Thanks!

**Comment by fredrik-telia** [comment]
> 

I might need some pointers, as there seems to be a test ( `parses characteristics from text tags` ) that passes, but shouldn't..

When using the player with a real stream, the `role` field in the stream object isn't filled, but the unit test seems to pass.

**Comment by fredrik-telia** [comment]
> 

I need some help understanding why this line in `hls_parser_unit.js` is truthy: `expect(actual).toEqual(manifest);` when they are, in fact, not equal at all

actual (using console.log(JSON.stringify(actual)):
```
'actual parsed: ', '{"presentationTimeline":{"presentationStartTime_":null,"presentationDelay_":0,"duration_":null,"segmentAvailabilityDuration_":null,"maxSegmentDuration_":1,"minSegmentStartTime_":null,"maxSegmentEndTime_":null,"clockOffset_":0,"static_":true,"userSeekStart_":0,"autoCorrectDrift_":true,"availabilityTimeOffset_":0,"startTimeLocked_":false},"variants":[{"id":2,"language":"und","primary":false,"audio":null,"video":{"id":1,"originalId":null,"segmentIndex":null,"mimeType":"video/mp4","codecs":"avc1","encrypted":false,"drmInfos":[],"keyIds":{},"language":"und","label":null,"type":"video","primary":false,"trickModeVideo":null,"emsgSchemeIdUris":null,"frameRate":60,"width":960,"height":540,"roles":[],"forced":false,"channelsCount":null,"audioSamplingRate":null,"spatialAudio":false,"closedCaptions":null},"bandwidth":200,"allowedByApplication":true,"allowedByKeySystem":true,"decodingInfos":[]}],"textStreams":[{"id":3,"originalId":"English (caption)","segmentIndex":null,"mimeType":"text/vtt","codecs":"","encrypted":false,"drmInfos":[],"keyIds":{},"language":"en","label":"English (caption)","type":"text","primary":true,"trickModeVideo":null,"emsgSchemeIdUris":null,"roles":[],"forced":false,"channelsCount":null,"audioSamplingRate":null,"spatialAudio":false,"closedCaptions":null},{"id":4,"originalId":"English (caption)","segmentIndex":null,"mimeType":"text/vtt","codecs":"","encrypted":false,"drmInfos":[],"keyIds":{},"language":"en","label":"English (caption)","type":"text","primary":true,"trickModeVideo":null,"emsgSchemeIdUris":null,"roles":[],"forced":false,"channelsCount":null,"audioSamplingRate":null,"spatialAudio":false,"closedCaptions":null}],"imageStreams":[],"offlineSessionIds":[],"minBufferTime":0,"sequenceMode":true}'
```

vs manifest (same deal using JSON.stringify):
```
'{"variants":[{"sample":{"audio":null,"video":{"sample":{"type":"video"}}}}],"textStreams":[{"sample":{"type":"text","language":"en","kind":"subtitle","mimeType":"application/mp4","codecs":""}},{"sample":{"type":"text","language":"en","kind":"subtitle","mimeType":"application/mp4","codecs":"","roles":["public.accessibility.describes-spoken-dialog","public.accessibility.describes-music-and-sound"]}}],"imageStreams":[],"presentationTimeline":{},"offlineSessionIds":[],"minBufferTime":0,"sequenceMode":true}'
```

Maybe I'm missing something?

**Comment by avelad** [comment]
@theodab can you review it? Thanks!

**Review by avelad — APPROVED** [review]

**Comment by martinstark** [comment]
@avelad the issue that @fredrik-telia is pointing out looks quite serious, since that would mean the unit tests that currently exist for hls parser don't actually verify anything. It looks like whatever two objects are compared will pass the test. Should this be reported as a separate issue?

**Review by martinstark — COMMENTED** [review]
would it make sense to also populate the `kind` property now that `roles` are being included?

**Review by joeyparrish — APPROVED** [review]

**Comment by joeyparrish** [comment]
> > 
> 
> I need some help understanding why this line in the function `testHlsParser` in `hls_parser_unit.js` is truthy:
> 
> `expect(actual).toEqual(manifest);` when they are, in fact, not equal at all
> 
> actual (using console.log(JSON.stringify(actual)):
> 
> ```
> '{"presentationTimeline":{"presentationStartTime_":null,"presentationDelay_":0,"duration_":null,"segmentAvailabilityDuration_":null,"maxSegmentDuration_":1,"minSegmentStartTime_":null,"maxSegmentEndTime_":null,"clockOffset_":0,"static_":true,"userSeekStart_":0,"autoCorrectDrift_":true,"availabilityTimeOffset_":0,"startTimeLocked_":false},"variants":[{"id":2,"language":"und","primary":false,"audio":null,"video":{"id":1,"originalId":null,"segmentIndex":null,"mimeType":"video/mp4","codecs":"avc1","encrypted":false,"drmInfos":[],"keyIds":{},"language":"und","label":null,"type":"video","primary":false,"trickModeVideo":null,"emsgSchemeIdUris":null,"frameRate":60,"width":960,"height":540,"roles":[],"forced":false,"channelsCount":null,"audioSamplingRate":null,"spatialAudio":false,"closedCaptions":null},"bandwidth":200,"allowedByApplication":true,"allowedByKeySystem":true,"decodingInfos":[]}],"textStreams":[{"id":3,"originalId":"English (caption)","segmentIndex":null,"mimeType":"text/vtt","codecs":"","encrypted":false,"drmInfos":[],"keyIds":{},"language":"en","label":"English (caption)","type":"text","primary":true,"trickModeVideo":null,"emsgSchemeIdUris":null,"roles":[],"forced":false,"channelsCount":null,"audioSamplingRate":null,"spatialAudio":false,"closedCaptions":null},{"id":4,"originalId":"English (caption)","segmentIndex":null,"mimeType":"text/vtt","codecs":"","encrypted":false,"drmInfos":[],"keyIds":{},"language":"en","label":"English (caption)","type":"text","primary":true,"trickModeVideo":null,"emsgSchemeIdUris":null,"roles":[],"forced":false,"channelsCount":null,"audioSamplingRate":null,"spatialAudio":false,"closedCaptions":null}],"imageStreams":[],"offlineSessionIds":[],"minBufferTime":0,"sequenceMode":true}'
> ```
> 
> vs manifest (same deal using JSON.stringify):
> 
> ```
> '{"variants":[{"sample":{"audio":null,"video":{"sample":{"type":"video"}}}}],"textStreams":[{"sample":{"type":"text","language":"en","kind":"subtitle","mimeType":"application/mp4","codecs":""}},{"sample":{"type":"text","language":"en","kind":"subtitle","mimeType":"application/mp4","codecs":"","roles":["public.accessibility.describes-spoken-dialog","public.accessibility.describes-music-and-sound"]}}],"imageStreams":[],"presentationTimeline":{},"offlineSessionIds":[],"minBufferTime":0,"sequenceMode":true}'
> ```
> 
> Maybe I'm missing something?

Here's an expanded version of what you logged in "manifest" for clarity:

```json
{
  "variants": [
    {
      "sample": {
        "audio": null,
        "video": {
          "sample": {
            "type": "video"
          }
        }
      }
    }
  ],
  "textStreams": [
    {
      "sample": {
        "type": "text",
        "language": "en",
        "kind": "subtitle",
        "mimeType": "application/mp4",
        "codecs": ""
      }
    },
    {
      "sample": {
        "type": "text",
        "language": "en",
        "kind": "subtitle",
        "mimeType": "application/mp4",
        "codecs": "",
        "roles": [
          "public.accessibility.describes-spoken-dialog",
          "public.accessibility.describes-music-and-sound"
        ]
      }
    }
  ],
  "imageStreams": [],
  "presentationTimeline": {},
  "offlineSessionIds": [],
  "minBufferTime": 0,
  "sequenceMode": true
}
```

What you're missing in the serialization through JSON is that some of those objects are not actual variants, but rather partial matchers constructed with `jasmine.objectContaining()`.  They only contain certain fields, and others will be ignored in the comparison.  This is intentional, and good to keep tests focused on the specific behaviors they are verifying.  When you serialize one of these partial matchers, you see it's `sample` field, which is a private field used by the matcher to drive it custom matching behavior within jasmine.

I'll double-check the existing unit tests to make sure this partial matching is working correctly.

**Comment by joeyparrish** [comment]
Partial matching is working, but the test has a particular step that loads the relevant data:

```js
await loadAllStreamsFor(actual);
```

This is unrealistic in this case, because it triggers lazy-loading for all streams.  The only way a user would trigger a similar code path is by cycling through every text track, audio language, and resolution.

So roles are only missing before lazy-loading, and every single test case triggers lazy-loading on every single track before checking expectations.

I'll add a test case that fails without your fix.

**Comment by joeyparrish** [comment]
> would it make sense to also populate the `kind` property now that `roles` are being included?

Looks like the answer is "yes".  That is also missing in the regression test I'm creating.

**Comment by joeyparrish** [comment]
I added a regression test.  You'll see that it's failing on two counts: missing `kind` and malformed `roles`.

**Comment by joeyparrish** [comment]
I fixed two bugs in this PR:
 - kind missing
 - roles should be an array, not a string

Both of these were caught by the new test I added.

File: lib/hls/hls_parser.js [hunk 1]
```diff
@@ -1949,9 +1947,7 @@ shaka.hls.HlsParser = class {
 
     const roles = [];
```
Comment (theodab) [inline lib/hls/hls_parser.js:1948]: This needs to no longer be a const

Comment (joeyparrish) [reply lib/hls/hls_parser.js]: Fixed

File: lib/hls/hls_parser.js [hunk 2]
```diff
@@ -1947,12 +1945,7 @@ shaka.hls.HlsParser = class {
     const kind = (type == shaka.util.ManifestParserUtils.ContentType.TEXT) ?
         shaka.util.ManifestParserUtils.TextStreamKind.SUBTITLE : undefined;
 
-    const roles = [];
-    if (characteristics) {
-      for (const characteristic of characteristics.split(',')) {
-        roles.push(characteristic);
-      }
-    }
+    const roles = characteristics ? characteristics.split(',') : [];
```
[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/hls/hls_parser.js]: Fixed.  Both roles and kind were redundant here after my other changes.

**Review by theodab — APPROVED** [review]

**Comment by joeyparrish** [comment]
Thank you to @fredrik-telia and @martinstark for highlighting the test issues.  I'm very glad to have both the bug and the test issues fixed now!

I will merge this when the test pass is complete.

**Review by avelad — APPROVED** [review]

**Pull request merged**